pubsublite/operations.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package pubsublite
+
+import (
+	"context"
+	"time"
+
+	vkit "cloud.google.com/go/pubsublite/apiv1"
+	pb "google.golang.org/genproto/googleapis/cloud/pubsublite/v1"
+	tspb "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// SeekTarget is the target location to seek a subscription to. Implemented by
+// BacklogLocation, PublishTime, EventTime.
+type SeekTarget interface {
+	setRequest(req *pb.SeekSubscriptionRequest)
+}
+
+// BacklogLocation refers to a location with respect to the message backlog.
+// It implements the SeekTarget interface.
+type BacklogLocation int
+
+const (
+	// End refers to the location past all currently published messages. End
+	// skips the entire message backlog.
+	End BacklogLocation = iota + 1
+
+	// Beginning refers to the location of the oldest retained message.
+	Beginning
+)
+
+func (b BacklogLocation) setRequest(req *pb.SeekSubscriptionRequest) {
```

Comment:
This is really cool! I didn't know Go allowed attaching behavior to types that are effectively ints.